llama-index-integrations/vector_stores/llama-index-vector-stores-redis/llama_index/vector_stores/redis/base.py

```diff
@@ -18,122 +18,110 @@
-                "{index_prefix}{prefix_ending}".
-            prefix_ending (str): Prefix ending for the index. Be careful when
-                changing this: https://github.com/jerryjliu/llama_index/pull/6665.
-                Defaults to "/vector".
-            index_args (Dict[str, Any]): Arguments for the index. Defaults to None.
-            metadata_fields (List[str]): List of metadata fields to store in the index
-                (only supports TAG fields).
-            redis_url (str): URL for the redis instance.
+            schema (IndexSchema, optional): Redis index schema object.
+            redis_client (Redis, optional): Redis client connection.
+            redis_url (str, optional): Redis server URL.
                 Defaults to "redis://localhost:6379".
-            overwrite (bool): Whether to overwrite the index if it already exists.
+            overwrite (bool, optional): Whether to overwrite the index if it already exists.
                 Defaults to False.
-            kwargs (Any): Additional arguments to pass to the redis client.
 
         Raises:
             ValueError: If redis-py is not installed
             ValueError: If RediSearch is not installed
 
-        Examples:
-            >>> from llama_index.core.vector_stores.redis import RedisVectorStore
-            >>> # Create a RedisVectorStore
-            >>> vector_store = RedisVectorStore(
-            >>>     index_name="my_index",
-            >>>     index_prefix="llama_index",
-            >>>     index_args={"algorithm": "HNSW", "m": 16, "ef_construction": 200,
-                "distance_metric": "cosine"},
-            >>>     redis_url="redis://localhost:6379/",
-            >>>     overwrite=True)
+        # TODO -- update example
         """
-        try:
-            # connect to redis from url
-            self._redis_client = redis.from_url(redis_url, **kwargs)
-            # check if redis has redisearch module installed
-            check_redis_modules_exist(self._redis_client)
-        except ValueError as e:
-            raise ValueError(f"Redis failed to connect: {e}")
-
-        # index identifiers
-        self._prefix = index_prefix + prefix_ending
-        self._index_name = index_name
-        self._index_args = index_args if index_args is not None else {}
-        self._metadata_fields = metadata_fields if metadata_fields is not None else []
+        # Setup schema
+        if not schema:
+            logger.info("Using default RedisVectorStore schema.")
+            schema = RedisVectorStoreSchema()
+
+        self._validate_schema(schema)
+
+        self._vector_field_name = next(
+            [name for name, field in schema.fields.items() if field.type == "vector"]
+        )
+        self._return_fields = [
+            self._node_id_field_name,
+            self._doc_id_field_name,
+            self._text_field_name,
+            self._text_field_name,
+            VectorQuery.DISTANCE_ID,
+            "_node_content",
```

Comment:
Does the schema have to contain all these field? Will it error if any of these aren't available? 

It might be nice if the user could use a vectored that wasn't built with llama-index. At a minimum, it would need a text and vector(?) field.

Reply:
So I was thinking through this a bit too.... the `return_fields` are not necessarily aligned to what gets "indexed" in Redis. These are just telling our API which fields to project as part of the response payload from the db.

However, for LlamaIndex to function properly, we do need a number of fields by default. I THINK we need atleast the `doc_id` (so we can do the deletion properly) and then some vector field to support KNN queries. The `text`, `node_id` (and any other) fields will be stored in Redis too, but can optionally be indexed if needed for a certain application.

With this in mind... the design I had in place for schema was to have a "default" schema with the minimum required fields.
- If user passes no schema, this would get used.
- If user passes their own schema, we do a quick validation to make sure it has the minimum required fields. (and fail clearly if there's an issue).

However this introduces one issue, where using the default schema... we have no way to know what the embedding dimension (or vector configuration) is until the user writes data, at which point we could amend the schema. But this is just a little bit prickly, doable though.

Reply:
This is the current set of fields I have marked as default (and required by any custom implementation) in the code:

```

class RedisVectorStoreSchema(IndexSchema):
    """The default Redis Vector Store Schema."""

    def __init__(self, **data) -> None:
        index = RedisIndexInfo()
        fields: Dict[str, BaseField] = {
            "id": FieldFactory.create_field("tag", "id", {"sortable": False}),
            "doc_id": FieldFactory.create_field("tag", "doc_id", {"sortable": False}),
            "text": FieldFactory.create_field("text", "text", {"weight": 1.0}),
        }
        super().__init__(index=index, fields=fields)

```

Reply:
Works for me! 👍🏻